File: Binary_Trees/116-populating-next-right-pointers-in-each-node/populating-next-right-pointers-in-each-node.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        if (root == nullptr) {
            return nullptr;
        }

        queue<Node*> temp;
        temp.push(root);

        while (!temp.empty()) {
            vector<Node*> lv_node;
            int size = temp.size();

            for (int i = 0; i < size; i++) {
                Node* x = temp.front();
                temp.pop();
                lv_node.push_back(x);

                if (x->left) {
                    temp.push(x->left);
                }
                if (x->right) {
                    temp.push(x->right);
                }
            }

            for (int i = 1; i < size; i++) {
                lv_node[i - 1]->next = lv_node[i];
            }
            lv_node[size - 1]->next = nullptr;
        }

        return root;
    }
};

File: Binary_Trees/116-populating-next-right-pointers-in-each-node/populating-next-right-pointers-in-each-node.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if root is None:
            return None
        
        temp = deque()

        temp.append(root)

        while len(temp) > 0:
            lv_node = []
            size = len(temp)

            for i in range(0, size):
                x = temp.popleft()
                lv_node.append(x)

                if x.left:
                    temp.append(x.left)

                if x.right:
                    temp.append(x.right)
            
            for i in range(1, size):
                lv_node[i-1].next = lv_node[i]
            
            lv_node[-1].next = None
        
        return root